common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationSupport.java

```diff
@@ -214,6 +214,207 @@ public static int execICU(final UTF8String string, final UTF8String substring,
     }
   }
 
+  public static class StringTrim {
+    public static UTF8String exec(
+        final UTF8String srcString,
+        final int collationId) {
+      CollationFactory.Collation collation = CollationFactory.fetchCollation(collationId);
+      if (collation.supportsBinaryEquality) {
+        return execBinary(srcString);
+      } else {
+        return execICU(srcString, collationId);
+      }
+    }
+    public static UTF8String exec(
+        final UTF8String srcString,
+        final UTF8String trimString,
+        final int collationId) {
+      CollationFactory.Collation collation = CollationFactory.fetchCollation(collationId);
+      if (collation.supportsBinaryEquality) {
+        return execBinary(srcString, trimString);
+      } else {
+        return execICU(srcString, trimString, collationId);
+      }
```

Comment:
why do we not have an `lcase` implementation?

Reply:
You requested that we don't have `lcase` separately in the older PR: https://github.com/apache/spark/pull/45749#discussion_r1546312983.

The logic is now embeded in `CollationAwareUTF8String.trimLeft` and `CollationAwareUTF8String.trimRight`.

Reply:
that was indeed the case, but now it would be best if we could pull off separating these things

- static classes for expressions in CollationSupport should handle "rounting" of binary/lowecase/ICU implementations
- CollationAwareUTF8String should "extend" UTF8String with appropriate collation awareness, but shouldn't branch based on collation

for example, `CollationAwareUTF8String.trimLeft` should no longer call `CollationFactory.fetchCollation(collationId).supportsLowercaseEquality`, but execLowercase implementation in the appropriate `CollationSupport` class should do trimString.toLowerCase() prior to calling it

Reply:
All right, will change accordingly and add separate methods to `CollationAwareUTF8String`.